pkg/controller/daemon/daemon_controller.go

```diff
@@ -1000,108 +999,98 @@ func (dsc *DaemonSetsController) syncNodes(ds *apps.DaemonSet, podsToDelete, nod
-					dsc.expectations.CreationObserved(dsKey)
-					errCh <- err
-					utilruntime.HandleError(err)
-				}
-			}(i)
+	var (
+		errs         []error
+		createsCount int
+	)
+
+	createBarrier := controller.NewTaskBarrier()
+	// Batch the pod creates and fail on first error.
+	createBarrier.Limiter = controller.NewSlowStarter()
+	createBarrier.Policy = controller.ErrorPolicyFunc(func(err error) bool {
+		// If we see a timeout, it's possible that the pod is created but its
+		// initialization has timed out. If the initialization is successful
+		// eventually, the controller will observe the creation via the informer.
+		// If the initialization fails, or if the pod keeps uninitialized for a
+		// long time, the informer will not receive any update, and the controller
+		// will create a new pod when the expectation expires.
+		//
+		// Hence, we ignore timeouts.
+		if err != nil && !errors.IsTimeout(err) {
+			klog.V(2).Infof("Failed creation, decrementing expectations for set %q/%q", ds.Namespace, ds.Name)
+			dsc.expectations.CreationObserved(dsKey)
+			errs = append(errs, err)
+			utilruntime.HandleError(err)
+			return false
 		}
-		createWait.Wait()
-		// any skipped pods that we never attempted to start shouldn't be expected.
-		skippedPods := createDiff - (batchSize + pos)
-		if errorCount < len(errCh) && skippedPods > 0 {
-			klog.V(2).Infof("Slow-start failure. Skipping creation of %d pods, decrementing expectations for set %q/%q", skippedPods, ds.Namespace, ds.Name)
-			for i := 0; i < skippedPods; i++ {
-				dsc.expectations.CreationObserved(dsKey)
+		return true
+	})
+
+	for i := 0; i < createDiff; i++ {
+		node := nodesNeedingDaemonPods[i]
+
+		createBarrier.Go(func() error {
+			podTemplate := template.DeepCopy()
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.ScheduleDaemonSetPods) {
+				// The pod's NodeAffinity will be updated to make sure the Pod is bound
+				// to the target node by default scheduler. It is safe to do so because there
+				// should be no conflicting node affinity with the target node.
+				podTemplate.Spec.Affinity = util.ReplaceDaemonSetPodNodeNameNodeAffinity(
+					podTemplate.Spec.Affinity, node)
+
+				return dsc.podControl.CreatePodsWithControllerRef(ds.Namespace, podTemplate,
+					ds, metav1.NewControllerRef(ds, controllerKind))
 			}
-			// The skipped pods will be retried later. The next controller resync will
-			// retry the slow start process.
-			break
+			// If pod is scheduled by DaemonSetController, set its '.spec.scheduleName'.
+			podTemplate.Spec.SchedulerName = "kubernetes.io/daemonset-controller"
+
+			return dsc.podControl.CreatePodsOnNode(node, ds.Namespace, podTemplate,
+				ds, metav1.NewControllerRef(ds, controllerKind))
+		})
+	}
+	createBarrier.Wait()
+
+	// any skipped pods that we never attempted to start shouldn't be expected.
+	if skippedPods := createDiff - createsCount; skippedPods > 0 {
```

Comment:
I might be missing something here, but seems like `createsCount` is never updated.

Reply:
Thanks, you are right. It should be incremented in createBarrier.Policy when no (or timeout) error is encountered. I'll add some tests around correctly updating expectations.